client/tinychain/ml/linalg.py

```diff
@@ -269,7 +273,464 @@ def step(i: UInt):
+            h = s*g
+            g = c*g
+
+            z = pythag(f,h)
+            e[i-1] = z
+
+            c = f/z
+            s = h/z
+            f = x*c+g*s
+            g = g*c-x*s
+            h = y*s
+            y = y*c
+
+            cxt.X = V[:,i-1].copy()
+            cxt.Z = V[:,i].copy()
+            V_upd = After([cxt.X, cxt.Z], V[:, i-1].write(c * cxt.X + s * cxt.Z))
+            V_upd_final = After(V_upd, V[:, i].write(c * cxt.Z - s * cxt.X))
+
+            z = pythag(f, h)
+            W[i-1] = z
+
+            @closure(c, s)
+            @post_op
+            def update_c_s(f: F32, z: F32, h: F32) -> Tuple:
+                c = f/z
+                s = h/z
+                return c, s
+
+            cxt.update_c_s = update_c_s(f, z, h)     
+            c, s = If(z >= EPS, cxt.update_c_s, Tuple([c, s]))
+
+            f = c*g+s*y
+            x = c*y-s*g
+
+            cxt.Y = U[:,i-1].copy()
+            cxt.T = U[:,i].copy()
+            U_upd = After([cxt.Y, cxt.T], U[:, i-1].write(c * cxt.Y + s * cxt.T))
+            U_upd_final = After(U_upd, U[:, i].write(c * cxt.T - s * cxt.Y))
+            return Map(U=U_upd_final, W=W, V=V_upd_final, e=e, f = f, x = x)
+
+        @closure(k)
+        @post_op
+        def cond(i: UInt) -> Bool:
+            return i < k+1
+
+        e[cxt.l] = 0.0
+        e[k] = f 
+        cxt.U_V_upd = Map(While(cond, step, Map(i=cxt.l+1, x = x, c = c, s = s, U=U, W=W, V=V, e=e)))
+        W_upd = Tensor(After(W_upd_init, cxt.U_V_upd['W']))
+        W_upd[k] = x
+
+        return Map(t=t + 1, l=cxt.l, W=W_upd, running=running, convergence=convergence, U=cxt.U_V_upd['U'], V=cxt.U_V_upd['V'], e=e)
+
+    @closure(maxiter)
+    @post_op
+    def cond(t: UInt, running: Bool, convergence: Bool) -> Bool:
+        return running.logical_and(t < maxiter).logical_and(convergence)
+
+    return Map(While(cond, step, Map(t=0, U=U, W=W, V=V, e=e)))
+
+
+@post_op
+def bidiagonalize(txn, A: Tensor):
+    """
+    Reduction to bidiagonal form
+    """    
+
+    U_init = A.copy()
+    txn.shape = U_init.shape
+    txn.m, txn.n = [UInt(dim) for dim in txn.shape.unpack(2)]
+    W_init = Dense.zeros((txn.n,), dtype=F32)
+    V_init = Dense.zeros((txn.n, txn.n), dtype=F32)
+    e_init = Dense.zeros((txn.n,), dtype=F32)
+
+    txn.householder = householder(U_init, W_init, e_init)
+    U_hh = txn.householder['U']
+    W_hh = txn.householder['W']
+    e_hh = txn.householder['e']
+    txn.rht = rht(U_hh, V_init, e_hh)
+    U_rht = txn.rht['U']
+    V_rht = txn.rht['V']
+    e_rht = txn.rht['e']
+    txn.lht = lht(U_rht, W_hh)
+    U_lht = txn.lht['U']
+    W_lht = txn.lht['W']
+
+    return Map(U=U_lht, W=W_lht, V=V_rht, e=e_rht)
+
+
+@post_op
+def svd(txn, matrix: Tensor, maxiter=30) -> Tuple:
     """Return the singular value decomposition of the given `matrix`."""
 
-    raise NotImplementedError("singular value decomposition")
+    # Bidiagonal form
+    txn.bidiagonalize = bidiagonalize(matrix)
```

Comment:
You should make a copy of the matrix before updating it, in case it's a view (which doesn't support writes)